### ISSUE TITLE
certbot Segfault Debian 9 GRSecurity

### ISSUE THREAD

**Issue by Pseudodevincent** [certbot/certbot#5621]
If you're having trouble using Certbot and aren't sure you've found a bug or
request for a new feature, please first try asking for help at
https://community.letsencrypt.org/. There is a much larger community there of
people familiar with the project who will be able to more quickly answer your
questions.

## My operating system is (include version):
Debian 9.3
4.9.0-4-grsec-amd64

## I installed Certbot with (certbot-auto, OS package manager, pip, etc):
sudo apt-get install python-certbot-apache

## I ran this command and it produced this output:
sudo certbot renew
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Processing /etc/letsencrypt/renewal/example1.com.conf
Cert not yet due for renewal
Processing /etc/letsencrypt/renewal/example2.com.conf
Cert not yet due for renewal
Processing /etc/letsencrypt/renewal/example3.com.conf
Cert not yet due for renewal
Processing /etc/letsencrypt/renewal/example4.com.conf
Cert is due for renewal, auto-renewing...
Erreur de segmentation

## Certbot's behavior differed from what I expected because:
It was working well before

## Here is a Certbot log showing the issue (if available):
From syslog:
Feb 24 21:43:30 ns330342 kernel: [26612.430473] grsec: From 81.249.218.220: denied RWX mmap of <anonymous mapping> by /usr/bin/certbot[certbot:30586] uid/euid:0/0 gid/egid:0/0, parent /usr/bin/sudo[sudo:30585] uid/euid:0/0 gid/egid:0/0
Feb 24 21:43:30 ns330342 kernel: [26612.442346] certbot[30586]: segfault at 0 ip 00000389df0e3c0b sp 000003c20e2899d8 error 6 in libffi.so.6.0.4[389df0de000+7000]
Feb 24 21:43:30 ns330342 kernel: [26612.451877] grsec: From 81.249.218.220: Segmentation fault occurred at            (nil) in /usr/bin/certbot[certbot:30586] uid/euid:0/0 gid/egid:0/0, parent /usr/bin/sudo[sudo:30585] uid/euid:0/0 gid/egid:0/0
Feb 24 21:43:31 ns330342 kernel: [26612.461546] grsec: From 81.249.218.220: denied resource overstep by requesting 4096 for RLIMIT_CORE against limit 0 for /usr/bin/certbot[certbot:30586] uid/euid:0/0 gid/egid:0/0, parent /usr/bin/sudo[sudo:30585] uid/euid:0/0 gid/egid:0/0
Feb 24 21:43:31 ns330342 kernel: [26612.511813] grsec: From 81.249.218.220: Segmentation fault occurred at 0000000000007779 in /usr/bin/sudo[sudo:30585] uid/euid:0/0 gid/egid:0/0, parent /bin/bash[bash:30563] uid/euid:1000/1000 gid/egid:1000/1000

**Comment by schoen** [comment]
Hi @Pseudodevincent,

Thanks for the bug report. This is probably a conflict between some library that we use and grsec's security policies, which might well turn out to be a bug in the libraries we use, or might require a grsec-specific configuration.

Could you also share the associated logs from `/var/log/letsencrypt` to try to show more specifically what Certbot was doing at the moment that it crashed?

**Comment by schoen** [comment]
Is it possible that this issue is a variant of #4996?

https://github.com/certbot/certbot/issues/4996

**Comment by Pseudodevincent** [comment]
Hello @schoen,

I used the workaround from @AlexanderCurl from issue #3295 and it solved my problem although I'm not very confortable with grsec and paxctld configuration. It's above my skills to understand the consequences in terms of security of what's in my /etc/paxctld.conf.
I just added the following line:
/usr/bin/certbot                m

Here is the /var/log/letsencrypt before using the workaround :
_2018-02-24 20:43:27,673:DEBUG:certbot.main:Root logging level set at 20
2018-02-24 20:43:27,677:INFO:certbot.main:Saving debug log to /var/log/letsencrypt/letsencrypt.log
2018-02-24 20:43:27,681:DEBUG:certbot.main:certbot version: 0.10.2
2018-02-24 20:43:27,681:DEBUG:certbot.main:Arguments: []
2018-02-24 20:43:27,685:DEBUG:certbot.main:Discovered plugins: PluginsRegistry(PluginEntryPoint#apache,PluginEntryPoint#webroot,PluginEntryPoint#null,PluginEntryPoint#manual,PluginEntryPoint#standalone)
2018-02-24 20:43:27,760:INFO:certbot.renewal:Cert not yet due for renewal
2018-02-24 20:43:27,775:INFO:certbot.renewal:Cert not yet due for renewal
2018-02-24 20:43:27,789:INFO:certbot.renewal:Cert not yet due for renewal
2018-02-24 20:43:27,805:DEBUG:certbot.storage:Should renew, less than 30 days before certificate expiry 2018-03-15 22:33:17 UTC.
2018-02-24 20:43:27,805:INFO:certbot.renewal:Cert is due for renewal, auto-renewing...
2018-02-24 20:43:27,807:DEBUG:certbot.plugins.selection:Requested authenticator apache and installer apache
2018-02-24 20:43:30,893:DEBUG:certbot.plugins.selection:Single candidate plugin: * apache
Description: Apache Web Server plugin - Beta
Interfaces: IAuthenticator, IInstaller, IPlugin
Entry point: apache = certbot_apache.configurator:ApacheConfigurator
Initialized: <certbot_apache.configurator.ApacheConfigurator object at 0x389da6053d0>
Prep: True
2018-02-24 20:43:30,899:DEBUG:certbot.plugins.selection:Single candidate plugin: * apache
Description: Apache Web Server plugin - Beta
Interfaces: IAuthenticator, IInstaller, IPlugin
Entry point: apache = certbot_apache.configurator:ApacheConfigurator
Initialized: <certbot_apache.configurator.ApacheConfigurator object at 0x389da6053d0>
Prep: True
2018-02-24 20:43:30,900:DEBUG:certbot.plugins.selection:Selected authenticator <certbot_apache.configurator.ApacheConfigurator object at 0x389da6053d0> and installer <certbot_apache.configurator.ApacheConfigurator object at 0x389da6053d0>
2018-02-24 20:43:30,913:DEBUG:certbot.main:Picked account: <Account(**************************)>
2018-02-24 20:43:30,918:DEBUG:root:Sending GET request to https://acme-v01.api.letsencrypt.org/directory.
2018-02-24 20:43:30,934:DEBUG:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): acme-v01.api.letsencrypt.org_
